csi/csi.go

```diff
@@ -257,18 +235,102 @@ func (s *OsdCsiServer) addEncryptionInfoToLabels(labels, csiSecrets map[string]s
 // Start is used to start the server.
 // It will return an error if the server is already running.
 func (s *OsdCsiServer) Start() error {
-	return s.GrpcServer.Start(func(grpcServer *grpc.Server) {
+	if err := s.GrpcServer.Start(func(grpcServer *grpc.Server) {
 		csi.RegisterIdentityServer(grpcServer, s)
 		csi.RegisterControllerServer(grpcServer, s)
 		csi.RegisterNodeServer(grpcServer, s)
-	})
+	}); err != nil {
+		return err
+	}
+
+	if s.config.Net == "unix" {
+		go func() {
+			err := autoSocketRecover(s)
+			if err != nil {
+				logrus.Errorf("failed to start CSI driver socket auto-recover watcher: %v", err)
+			}
+		}()
+	}
+
+	return nil
 }
 
 // Start is used to stop the server.
 func (s *OsdCsiServer) Stop() {
 	s.GrpcServer.Stop()
 }
 
+func createGrpcServer(config *OsdCsiServerConfig) (*grpcserver.GrpcServer, error) {
+	// create correlation interceptor
+	var unaryInterceptors []grpc.UnaryServerInterceptor
+	correlationInterceptor := correlation.ContextInterceptor{
+		Origin: correlation.ComponentCSIDriver,
+	}
+	opts := make([]grpc.ServerOption, 0)
+	unaryInterceptors = append(unaryInterceptors, correlationInterceptor.ContextUnaryServerInterceptor)
+
+	// create scheduler interceptor
+	switch config.SchedulerName {
+	case "kubernetes":
+		logrus.Infof("CSI K8s filter being added for %s scheduler", config.SchedulerName)
+		ki := k8s.NewInterceptor()
+		unaryInterceptors = append(unaryInterceptors, ki.SchedUnaryInterceptor)
+
+	default:
+		logrus.Infof("No CSI filter being added for %s scheduler", config.SchedulerName)
+	}
+
+	// Add interceptors
+	opts = append(opts, grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(unaryInterceptors...)))
+
+	// Create server
+	gServer, err := grpcserver.New(&grpcserver.GrpcServerConfig{
+		Name:    "CSI 1.7",
+		Net:     config.Net,
+		Address: config.Address,
+		Opts:    opts,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create CSI server: %v", err)
+	}
+
+	return gServer, nil
+}
+
+func autoSocketRecover(s *OsdCsiServer) error {
+	socketPath := strings.TrimPrefix(s.Address(), "unix://")
+
+	// Start checking for CSI socket delete
+	for {
+		time.Sleep(csiSocketCheckInterval)
+
+		// Check if socket deleted
+		_, err := os.Stat(socketPath)
+		if err == nil {
+			continue
+		}
+
+		logrus.Infof("Detected CSI socket deleted at path %s. Stopping CSI server", socketPath)
+		s.Stop()
```

Comment:
Even though the socket path was deleted, the process is still running. We need to stop the server and re-create it.

s.Stop() handles killing the net.Listener goroutine.